File: go/scheduler.go
package main

import (
	"context"
	"fmt"
	"net/http"
	"sync"
	"time"
)

// Scheduler runs one goroutine per monitor. A buffered channel acts as a
// semaphore to bound concurrent in-flight HTTP checks (Section 4.2).
//
// TODO: persist CheckResult rows instead of printing
// TODO: structured JSON logging via log/slog instead of fmt.Printf
type Scheduler struct {
	sem      chan struct{}
	client   *http.Client
	cancels  map[string]context.CancelFunc
	mu       sync.Mutex
	wg       sync.WaitGroup
}

func NewScheduler(maxConcurrentChecks int) *Scheduler {
	return &Scheduler{
		sem:     make(chan struct{}, maxConcurrentChecks),
		client:  &http.Client{},
		cancels: make(map[string]context.CancelFunc),
	}
}

func (s *Scheduler) Register(m *Monitor) {
	ctx, cancel := context.WithCancel(context.Background())
	s.mu.Lock()
	s.cancels[m.ID] = cancel
	s.mu.Unlock()

	s.wg.Add(1)
	go s.runLoop(ctx, m)
}

func (s *Scheduler) runLoop(ctx context.Context, m *Monitor) {
	defer s.wg.Done()
	ticker := time.NewTicker(time.Duration(m.IntervalSeconds) * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return // unregistered or shutting down
		case <-ticker.C:
			s.sem <- struct{}{} // acquire
			s.performCheck(ctx, m)
			<-s.sem // release
		}
	}
}

func (s *Scheduler) performCheck(parent context.Context, m *Monitor) {
	ctx, cancel := context.WithTimeout(parent, time.Duration(m.TimeoutMs)*time.Millisecond)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, m.URL, nil)
	if err != nil {
		fmt.Printf(`{"monitor_id":%q,"error":%q}`+"\n", m.ID, err.Error())
		return
	}

	start := time.Now()
	resp, err := s.client.Do(req)
	latency := time.Since(start).Milliseconds()

	if err != nil {
		// TODO: distinguish ctx.DeadlineExceeded (timeout) vs connection refused, etc.
		fmt.Printf(`{"monitor_id":%q,"error":%q}`+"\n", m.ID, err.Error())
		return
	}
	defer resp.Body.Close()

	success := resp.StatusCode < 400
	// TODO: call m.RecordResult(success) and update status per Section 4
	fmt.Printf(`{"monitor_id":%q,"status_code":%d,"latency_ms":%d,"success":%v}`+"\n",
		m.ID, resp.StatusCode, latency, success)
}

func (s *Scheduler) Unregister(monitorID string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if cancel, ok := s.cancels[monitorID]; ok {
		cancel()
		delete(s.cancels, monitorID)
	}
}

// Shutdown cancels all monitor loops and waits (with a grace period) for
// in-flight checks to finish. Call from a SIGTERM handler.
func (s *Scheduler) Shutdown(ctx context.Context) {
	s.mu.Lock()
	for _, cancel := range s.cancels {
		cancel()
	}
	s.mu.Unlock()

	done := make(chan struct{})
	go func() {
		s.wg.Wait()
		close(done)
	}()

	select {
	case <-done:
	case <-ctx.Done():
		// TODO: log "forced shutdown, some checks may not have finished"
	}
}

File: python/scheduler.py
import asyncio
import time
import httpx
import structlog
from models import Monitor

log = structlog.get_logger()


class Scheduler:
    """
    One asyncio.Task per monitor, looping on its own interval.
    asyncio.Semaphore bounds concurrent in-flight HTTP checks (Section 4.2).

    TODO: persist CheckResult rows instead of just logging
    TODO: graceful shutdown — cancel all tasks, await them with a grace period
    """

    def __init__(self, max_concurrent_checks: int):
        self._semaphore = asyncio.Semaphore(max_concurrent_checks)
        self._tasks: dict[str, asyncio.Task] = {}
        self._client = httpx.AsyncClient()

    def register(self, monitor: Monitor) -> None:
        task = asyncio.create_task(self._run_loop(monitor))
        self._tasks[monitor.id] = task

    async def _run_loop(self, monitor: Monitor) -> None:
        try:
            while True:
                await asyncio.sleep(monitor.interval_seconds)
                async with self._semaphore:  # acquire/release bounded concurrency
                    await self._perform_check(monitor)
        except asyncio.CancelledError:
            # expected on unregister/shutdown — don't let it crash the scheduler
            raise

    async def _perform_check(self, monitor: Monitor) -> None:
        start = time.monotonic()
        try:
            resp = await asyncio.wait_for(
                self._client.get(monitor.url),
                timeout=monitor.timeout_ms / 1000,
            )
            latency_ms = int((time.monotonic() - start) * 1000)
            success = resp.status_code < 400
            # TODO: monitor.record_result(success) per Section 4 status rules
            await log.ainfo(
                "check_completed",
                monitor_id=monitor.id,
                status_code=resp.status_code,
                latency_ms=latency_ms,
                success=success,
            )
        except asyncio.TimeoutError:
            # TODO: record CheckResult with error="timeout"
            await log.awarning("check_timeout", monitor_id=monitor.id)
        except httpx.HTTPError as e:
            # TODO: record CheckResult with error=str(e)
            await log.awarning("check_failed", monitor_id=monitor.id, error=str(e))

    def unregister(self, monitor_id: str) -> None:
        task = self._tasks.pop(monitor_id, None)
        if task:
            task.cancel()

    async def shutdown(self) -> None:
        for task in self._tasks.values():
            task.cancel()
        await asyncio.gather(*self._tasks.values(), return_exceptions=True)
        await self._client.aclose()
